File: immutable.cpp
#include <iostream>
#include <vector>
#include <string>

std::vector<int> qsort(const std::vector<int>& xs) {
    if (xs.size() <= 1) {
        return xs;
    }

    int pivot = xs[0];
    std::vector<int> left;
    std::vector<int> right;
    left.reserve(xs.size());
    right.reserve(xs.size());

    for (size_t i = 1; i < xs.size(); ++i) {
        if (xs[i] < pivot) {
            left.push_back(xs[i]);
        } else {
            right.push_back(xs[i]);
        }
    }

    std::vector<int> sorted_left = qsort(left);
    std::vector<int> sorted_right = qsort(right);

    sorted_left.reserve(sorted_left.size() + 1 + sorted_right.size());
    sorted_left.push_back(pivot);
    sorted_left.insert(sorted_left.end(), sorted_right.begin(), sorted_right.end());
    return sorted_left;
}

int main(int argc, char* argv[]) {
    bool verbose = false;
    bool prePrint = false;
    for (int i = 1; i < argc; ++i) {
        if (std::string(argv[i]) == "-v") {
            verbose = true;
        } else if (std::string(argv[i]) == "-p") {
            prePrint = true;
        }
    }
    std::vector<int> xs(10000);
    for (int i = 0; i < 10000; ++i) {
        xs[i] = (i * 73 + 19) % 10000;
    }
    if (prePrint) {
        std::cout << "Before sorting:\n";
        for (int v : xs) {
            std::cout << v << " ";
        }
        std::cout << "\n";
        return 0;
    }
    auto ys = qsort(xs);
    if (verbose) {
        for (int v : ys) {
            std::cout << v << " ";
        }
        std::cout << "\n";
    } else {
        for (int i = 0; i < 10; ++i) {
            std::cout << ys[i] << " ";
        }
        std::cout << "\n";
    }
}

File: immutable.rs
use std::env;

fn quicksort(arr: &[i32]) -> Vec<i32> {
    if arr.len() <= 1 {
        return arr.to_vec();
    }
    
    let pivot = arr[0];
    let mut left = Vec::new();
    let mut right = Vec::new();
    
    // Pre-allocate capacity for better performance
    left.reserve(arr.len() / 2);
    right.reserve(arr.len() / 2);
    
    for &x in &arr[1..] {
        if x < pivot {
            left.push(x);
        } else {
            right.push(x);
        }
    }
    
    let mut sorted_left = quicksort(&left);
    sorted_left.push(pivot);
    sorted_left.extend(quicksort(&right));
    sorted_left
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let verbose = args.contains(&String::from("-v"));
    let pre_print = args.contains(&String::from("-p"));
    
    // Generate test data
    let mut arr = Vec::with_capacity(10000);
    for i in 0..10000 {
        arr.push((i * 73 + 19) % 10000);
    }
    
    if pre_print {
        println!("Before sorting:");
        println!("{:?}", arr);
        return;
    }
    
    let sorted = quicksort(&arr);
    
    if verbose {
        println!("{:?}", sorted);
    } else {
        println!("{:?}", &sorted[..10]);
    }
}
